java-tools/OsmAndServerUtilities/src/main/java/net/osmand/mailsender/EmailSenderTemplate.java

```diff
@@ -0,0 +1,470 @@
+	}
+
+	// configure by environment
+	public EmailSenderTemplate() {
+		final String templates = System.getenv("EMAIL_TEMPLATES");
+		if (templates != null) {
+			this.defaultTemplatesDirectory = templates;
+			LOG.info("Using env EMAIL_TEMPLATES: " + templates);
+		}
+
+		final String smtpServer = System.getenv("SMTP_SERVER");
+		if (smtpServer != null) {
+			LOG.info("Using env SMTP_SERVER: " + smtpServer);
+		}
+
+		final String apiKey = System.getenv("SENDGRID_KEY");
+		if (apiKey != null) {
+			LOG.info("Using env SENDGRID_KEY: qwerty :-)");
+		}
+
+		testEmailCopy = System.getenv("TEST_EMAIL_COPY");
+
+		this.sender = new SmtpSendGridSender(smtpServer, apiKey);
+	}
+
+	// send out email(s)
+	public EmailSenderTemplate send() throws UnsupportedEncodingException {
+		validateLoadedTemplates(); // final validation before send
+
+		for (String to : toList) {
+			setVarsByTo(to);
+
+			// build SendGrid-compatible objects
+
+			Email toObject = new Email(to);
+
+			Email fromObject = new Email(fill(fromEmail));
+			fromObject.setName(fill(fromName));
+
+			Content contentObject = new Content("text/html", fill(body));
+
+			Mail mailObject = new Mail(fromObject, fill(subject), toObject, contentObject);
+			mailObject.from = fromObject;
+
+			try {
+				Response response = sender.send(mailObject);
+				LOG.info(to.replaceFirst(".....", ".....") + ": sender response code: " + response.getStatusCode());
+				if (response.getStatusCode() == sender.STATUS_CODE_SENT) {
+					sentEmails++;
+				}
+			} catch (Exception e) {
+				LOG.warn(e.getMessage(), e);
+			}
+		}
+		return this;
+	}
+
+	// load template from file by template name and language code
+	public EmailSenderTemplate load(String template, @Nullable String langNullable) throws FileNotFoundException {
+		String lang = langNullable == null ? "en" : langNullable;
+		include("defaults", lang, false); // settings (email-headers, vars, etc)
+		include("header", lang, false); // optional
+		include(template, lang, true); // template required
+		include("footer", lang, false); // optional
+		include("unsubscribe", lang, false); // optional
+		return this;
+	}
+
+	// load template from file (using default lang)
+	public EmailSenderTemplate load(String template) throws FileNotFoundException {
+		return load(template, "en");
+	}
+
+	// load template from string
+	public EmailSenderTemplate template(String templateAsString) {
+		parse(Arrays.asList(templateAsString.split("\n")));
+		return this;
+	}
+
+	// load template from list
+	public EmailSenderTemplate template(List<String> templateAsList) {
+		parse(templateAsList);
+		return this;
+	}
+
+	// optional set fromEmail, use before load()
+	public EmailSenderTemplate from(String from) {
+		fromEmail = from;
+		return this;
+	}
+
+	// optional set fromName, use before load()
+	public EmailSenderTemplate name(String name) {
+		fromName = name;
+		return this;
+	}
+
+	// optional set subject, use before load()
```

Comment:
Comments to explain code bad practice, it's better to use design patterns / approaches and self explaining code. The best API that you can't use incorrectly

Reply:
comments removed, documentation clarified